tsdb/shard.go

```diff
@@ -1959,48 +2064,173 @@ func (fs *MeasurementFieldSet) marshalMeasurementFieldSetNoLock() (marshalled []
+			}
+			set := &MeasurementFields{}
+			set.fields.Store(fields)
+			fs.fields[string(measurement.GetName())] = set
+		}
+		return nil
+	}()
+
+	if err != nil {
+		return fmt.Errorf("failed loading %s: %w", fs.path, err)
+	}
+	return fs.ApplyChanges()
+}
+
+func (fs *MeasurementFieldSet) loadFieldIndex() (pb *internal.MeasurementFieldSet, rErr error) {
+	pb = &internal.MeasurementFieldSet{}
 
 	fd, err := os.Open(fs.path)
 	if os.IsNotExist(err) {
-		return nil
+		return pb, nil
 	} else if err != nil {
-		return err
+		return nil, err
 	}
-	defer fd.Close()
+
+	defer errors2.Capture(&rErr, fd.Close)()
 
 	var magic [4]byte
 	if _, err := fd.Read(magic[:]); err != nil {
-		return err
+		return nil, err
 	}
 
 	if !bytes.Equal(magic[:], fieldsIndexMagicNumber) {
-		return ErrUnknownFieldsFormat
+		return nil, fmt.Errorf("%q: %w", fs.path, ErrUnknownFieldsFormat)
 	}
 
-	var pb internal.MeasurementFieldSet
 	b, err := io.ReadAll(fd)
 	if err != nil {
-		return err
+		return nil, err
+	}
+	return pb, proto.Unmarshal(b, pb)
+}
+
+func (fscm *measurementFieldSetChangeMgr) loadAllFieldChanges() (changes []FieldChanges, rErr error) {
+	var fcs FieldChanges
+
+	fscm.mu.Lock()
+	defer fscm.mu.Unlock()
+	fd, err := os.Open(fscm.Path)
+	if os.IsNotExist(err) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	defer errors2.Capture(&rErr, fd.Close)()
+	for fcs, err = fscm.loadFieldChangeSet(fd); err == nil; fcs, err = fscm.loadFieldChangeSet(fd) {
+		changes = append(changes, fcs)
 	}
+	if errors.Is(err, io.EOF) {
+		return changes, nil
+	}
+	return nil, err
+}
+
+func (fscm *measurementFieldSetChangeMgr) loadFieldChangeSet(fd *os.File) (FieldChanges, error) {
+	var pb internal.FieldChangeSet
 
+	b, err := readSizePlusBuffer(fd, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed reading %s: %w", fd.Name(), err)
+	}
 	if err := proto.Unmarshal(b, &pb); err != nil {
+		return nil, fmt.Errorf("failed unmarshalling %s: %w", fd.Name(), err)
+	}
+
+	fcs := make([]*FieldChange, 0, len(pb.Changes))
+
+	for _, fc := range pb.Changes {
+		fcs = append(fcs, &FieldChange{
+			FieldCreate: FieldCreate{
+				Measurement: fc.Measurement,
+				Field: &Field{
+					ID:   0,
+					Name: string(fc.Field.Name),
+					Type: influxql.DataType(fc.Field.Type),
+				},
+			},
+			Delete: fc.Delete,
+		})
+	}
+	return fcs, nil
+}
+
+func (fs *MeasurementFieldSet) ApplyChanges() error {
+	changes, err := fs.changeMgr.loadAllFieldChanges()
+	if err != nil {
 		return err
 	}
+	if len(changes) <= 0 {
+		return os.RemoveAll(fs.changeMgr.Path)
+	}
 
-	fs.fields = make(map[string]*MeasurementFields, len(pb.GetMeasurements()))
-	for _, measurement := range pb.GetMeasurements() {
-		fields := make(map[string]*Field, len(measurement.GetFields()))
-		for _, field := range measurement.GetFields() {
-			fields[string(field.GetName())] = &Field{Name: string(field.GetName()), Type: influxql.DataType(field.GetType())}
+	for _, fcs := range changes {
```

Comment:
My preference is to leave the initial application of the changes untouched, and simply record them for later re-application on failure.  The `ApplyChanges` code will never be called during correct operation (I think the call in `Close` is erroneous, and will be removed.

Basically, changes are made as before, but a change log is written out, rather than the whole `MeasurementFieldSet`.  On `Close()`, the whole file is written out, and the change log is deleted. The change log is only applied on startup when it is found, indicating an unclean shutdown.